### DISCUSSION TITLE
Do we have any methods that can extend custom element tags?

### DISCUSSION THREAD

**Discussion by cyz1901** [armanbilge/calico#320] [General]
Hi @armanbilge. I really love Calico :kissing_heart:.  And after writing a demo, I think it's a good way to learn functional programming. 
I am trying to combine Calico and the JavaScript ecosystem in the mobile product application to create more possibilities. 

The type we use in DSL seems to be implemented based on the `fs2.dom` of the wrapper  `scala.js.dom`

Usually, this is sufficient.
But when I combine some libraries, there are custom tags in the library.
For example, the `nav` in https://tw-elements.com/docs/standard/navigation/navbar/

```
  .evalTap(
      _.innerHtml.set(
        """
              <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" class="inline-block w-5 h-5 stroke-current">
                <path id="heart-path" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 6h16M4 12h16M4 18h16">
                </path>
              </svg>
        """
      )
    )
```
But this not only looks ugly but also to some extent loses flexibility.

Do we provide other ways to extend new tags and write them in DSL?

like
`lazy val nav: HtmlTag[F, HtmlElement[F]] = htmlTag("nav")`

**Comment by armanbilge** [comment]
We have `nav`, it's actually called `navTag`.

https://www.javadoc.io/doc/com.armanbilge/calico-docs_sjs1_3/latest/calico/html/Html.html#navTag-0

We generate the elements using the scala-dom-types library. You can check there:
https://github.com/raquo/scala-dom-types/blob/04e66590f96216fe1e20036c4d9a32d59205526b/shared/src/main/scala/com/raquo/domtypes/defs/tags/MiscTagDefs.scala#L64-L78

**Comment by cyz1901** [comment]
Hey @armanbilge 
What is our version distribution strategy？
I noticed that some commits have updated dependent versions, but it seems that these versions have not been released.

I added [something](https://github.com/raquo/scala-dom-types/pull/100) for scala-dom-types

If merged, we will have a new version of scala-dom-types.

Should I create a new PR for Calico to upgrade the dependent version and release the new version for use in other projects?

**Reply by armanbilge** [reply]
> Should I create a new PR for Calico to upgrade the dependent version

Has it been released yet? We get update PRs automatically as soon as new versions are released.

> release the new version for use

I can cut a release after that update.

**Reply by armanbilge** [reply]
https://github.com/armanbilge/calico/releases/tag/v0.2.2

**Comment by cyz1901** [comment]
> > Should I create a new PR for Calico to upgrade the dependent version
> 
> Has it been released yet? We get update PRs automatically as soon as new versions are released.
> 
> > release the new version for use
> 
> I can cut a release after that update.

I already megre pr in scala-dom-types. And scala-dom-types make a release later this week. 
Is there any other way for us to add custom attributes directly in calico?  This way I can skip the tedious publishing process.

**Reply by armanbilge** [reply]
In Calico you always have access to the underlying DOM element. So anything is possible, e.g.:
```scala
img().evalTap { i =>
  IO {
    i.asInstanceOf[js.Dynamic].crossOrigin = "use-credentials"
  }
}
```

With a little effort we can add some helpers to make it nicer. But your SDT changes will be released soon hopefully :)

**Reply by armanbilge** [reply]
> But your SDT changes will be released soon hopefully :)

@cyz1901 https://github.com/armanbilge/calico/releases/tag/v0.2.2

**Reply by cyz1901** [reply]
The new version is amazing!  lol
